src/transformers/generation/tf_logits_process.py

```diff
@@ -585,3 +586,137 @@ def _force_token(generation_idx):
+        if generate_config.forced_decoder_ids[-1][1] == self.no_timestamps_token_id:
+            self.begin_index -= 1
+        self.max_initial_timestamp_index = generate_config.max_initial_timestamp_index
+
+    def __call__(self, input_ids, scores, cur_len):
+        def _un_force_token(current_token):
+            # scores[:,current_tokens ] = -float("inf")
+            batch_size = scores.shape[0]
+            new_scores = tf.identity(scores)
+            indices = tf.stack((tf.range(batch_size), tf.tile([current_token], [batch_size])), axis=1)
+            updates = tf.ones((batch_size,), dtype=scores.dtype) * -float("inf")
+            new_scores = tf.tensor_scatter_nd_update(new_scores, indices, updates)
+            return new_scores
+
+        def _force_tokens(current_tokens):
+            # scores[:, :] = -float("inf")
+            # scores[:,current_tokens ] = 0
+            batch_size = scores.shape[0]
+            new_scores = tf.ones_like(scores, dtype=scores.dtype) * -float("inf")
+            indices = tf.stack((tf.range(batch_size), tf.tile([current_tokens], [batch_size])), axis=1)
+            updates = tf.zeros((batch_size,), dtype=scores.dtype)
+            new_scores = tf.tensor_scatter_nd_update(new_scores, indices, updates)
+            return new_scores
+
+        # suppress <|notimestamps|> which is handled by without_timestamps
+        scores = _un_force_token(self.no_timestamps_token_id)  # scores[:, self.no_timestamps_token_id] = -float("inf")
+        scores = tf.cond(
+            tf.equal(cur_len, self.begin_index - 1),
+            lambda: _force_tokens(self.timestamp_begin),
+            lambda: tf.identity(scores),
+        )
+        if cur_len == self.begin_index - 1:
+            return scores
+
+        not_eos = (
+            tf.math.count_nonzero(((input_ids[:, self.begin_index : cur_len + 1] != self.eos_token_id)), axis=-1) >= 1
+        )
+        last_was_timestamps = not_eos & (input_ids[:, cur_len - 1] >= self.timestamp_begin)
+
+        not_eos = (
+            tf.math.count_nonzero((input_ids[:, self.begin_index : cur_len + 1] != self.eos_token_id), axis=-1) < 2
+        )
+        penultimate_was_timestamp = not_eos | (input_ids[:, cur_len - 2] >= self.timestamp_begin)
+
+        # timestamps have to appear in pairs, except directly before eos_token; mask logits accordingly
+        # for k in range(input_ids.shape[0]):
+        #     seq = [t for t in input_ids[k, self.begin_index :].tolist()]
+        #     last_was_timestamp = len(seq) >= 1 and seq[-1] >= self.timestamp_begin
+        #     penultimate_was_timestamp = len(seq) < 2 or seq[-2] >= self.timestamp_begin
+        #     if last_was_timestamp:
+        #          if penultimate_was_timestamp:  # has to be non-timestamp
+        #               scores[k, self.timestamp_begin :] = -float("inf")
+        #          else:  # cannot be normal text tokens
+        #               scores[k, : self.eos_token_id] = -float("inf")
```

Comment:
This part seems to be easier to solve with masking, rather than with a loop-based update 🤔 It will also be much faster, assuming it works 🐎 

Here's an approximation, based on the commented code
```py
# len(seq) == 1 corresponds to cur_len == self.begin_index + 2
last_was_timestamp = (cur_len >= self.begin_index + 2) & (input_ids[:, cur_len - 1] >= self.timestamp_begin)
penultimate_was_timestamp = (cur_len < self.begin_index + 3) | (input_ids[:, cur_len - 2] >= self.timestamp_begin)

# build mask for scores[k, self.timestamp_begin :] = -float("inf") when last_was_timestamp and penultimate_was_timestamp
row_mask_cannot_be_timestamp = last_was_timestamp & penultimate_was_timestamp
col_mask_cannot_be_timestamp = tf.concat((tf.zeros((self.timestamp_begin,)), tf.ones((scores.shape[1] - self.timestamp_begin,))), axis=0)
mask_cannot_be_timestamp = tf.broadcast_to(tf.expand_dims(row_mask_cannot_be_timestamp, axis=1), scores.shape) & tf.broadcast_to(tf.expand_dims(col_mask_cannot_be_timestamp, axis=0), scores.shape)

# build mask for scores[k, : self.eos_token_id] = -float("inf") when last_was_timestamp and not penultimate_was_timestamp
row_mask_cannot_be_text= last_was_timestamp & ~penultimate_was_timestamp
col_mask_cannot_be_text = tf.concat((tf.ones((self.eos_token_id,)), tf.zeros((scores.shape[1] - self.eos_token_id,))), axis=0)
mask_cannot_be_text = tf.broadcast_to(tf.expand_dims(row_mask_cannot_be_text, axis=1), scores.shape) & tf.broadcast_to(tf.expand_dims(col_mask_cannot_be_text, axis=0), scores.shape)

# merge masks and apply them
mask = mask_cannot_be_timestamp | mask_cannot_be_text
scores = tf.where(condition=mask, x=-float("inf"), y=scores)
```

Let me know if it helps and if it works :p

Reply:
Note that `col_mask_cannot_be_timestamp` and `col_mask_cannot_be_text` can be initialized at `__init__` time, for max speed 🚤

Reply:
works, you TF genius

Reply:
🤗